File: bindings/go/invalidation.go
package forge

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
)

const (
	InvalidationSchemaVersion = 1
	MaxInvalidationBytes      = 4096
)

// InvalidationEvent is a bounded, transport-neutral hint. It carries no authoritative state.
type InvalidationEvent struct {
	SchemaVersion uint32   `json:"schema_version"`
	Tags          []string `json:"tags"`
	QueryKeys     [][]any  `json:"query_keys"`
	Revision      *string  `json:"revision,omitempty"`
}

// EncodeInvalidationEvent validates and normalizes a version-1 hint.
func EncodeInvalidationEvent(event InvalidationEvent) ([]byte, error) {
	normalized, err := normalizeInvalidationEvent(event)
	if err != nil {
		return nil, err
	}
	encoded, err := json.Marshal(normalized)
	if err != nil {
		return nil, forgeError(CodeInvalid, "invalidation.encode", "invalidation event must contain JSON values")
	}
	if len(encoded) > MaxInvalidationBytes {
		return nil, forgeError(CodeLimit, "invalidation.encode", "invalidation event exceeds 4096 bytes")
	}
	return encoded, nil
}

// DecodeInvalidationEvent decodes a bounded hint and ignores unknown additive version-1 fields.
func DecodeInvalidationEvent(encoded []byte) (InvalidationEvent, error) {
	if len(encoded) > MaxInvalidationBytes {
		return InvalidationEvent{}, forgeError(CodeLimit, "invalidation.decode", "invalidation event exceeds 4096 bytes")
	}
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.UseNumber()
	var event InvalidationEvent
	if err := decoder.Decode(&event); err != nil {
		return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.decode", "invalidation event must be valid JSON")
	}
	if err := requireJSONEOF(decoder); err != nil {
		return InvalidationEvent{}, err
	}
	return normalizeInvalidationEvent(event)
}

func requireJSONEOF(decoder *json.Decoder) error {
	var trailing any
	err := decoder.Decode(&trailing)
	if errors.Is(err, io.EOF) {
		return nil
	}
	return forgeError(CodeInvalid, "invalidation.decode", "invalidation event must contain one JSON value")
}

func normalizeInvalidationEvent(event InvalidationEvent) (InvalidationEvent, error) {
	if event.SchemaVersion != InvalidationSchemaVersion {
		return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "unsupported invalidation schema version")
	}
	if len(event.Tags) == 0 && len(event.QueryKeys) == 0 {
		return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "invalidation event requires a target")
	}
	if len(event.Tags) > 32 || len(event.QueryKeys) > 32 || len(event.Tags)+len(event.QueryKeys) > 64 {
		return InvalidationEvent{}, forgeError(CodeLimit, "invalidation.validate", "invalidation event has too many targets")
	}
	seen := make(map[string]struct{}, len(event.Tags))
	for _, tag := range event.Tags {
		if len(tag) == 0 || len([]byte(tag)) > 128 {
			return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "invalidation tags must be 1..=128 UTF-8 bytes")
		}
		if _, duplicate := seen[tag]; duplicate {
			return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "invalidation tags must be unique")
		}
		seen[tag] = struct{}{}
	}
	for _, queryKey := range event.QueryKeys {
		if len(queryKey) == 0 || len(queryKey) > 8 {
			return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "query-key fragments must contain 1..=8 parts")
		}
		nodes := 0
		for _, part := range queryKey {
			encoded, err := json.Marshal(part)
			if err != nil {
				return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "query-key parts must be JSON values")
			}
			decoder := json.NewDecoder(bytes.NewReader(encoded))
			decoder.UseNumber()
			var normalized any
			if err := decoder.Decode(&normalized); err != nil {
				return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "query-key parts must be JSON values")
			}
			if err := validateInvalidationValue(normalized, 1, &nodes); err != nil {
				return InvalidationEvent{}, err
			}
		}
	}
	if event.Revision != nil && (len(*event.Revision) == 0 || len([]byte(*event.Revision)) > 256) {
		return InvalidationEvent{}, forgeError(CodeInvalid, "invalidation.validate", "invalidation revision must be 1..=256 UTF-8 bytes")
	}
	event.Tags = append([]string(nil), event.Tags...)
	event.QueryKeys = append([][]any(nil), event.QueryKeys...)
	encoded, err := json.Marshal(event)
	if err != nil || len(encoded) > MaxInvalidationBytes {
		return InvalidationEvent{}, forgeError(CodeLimit, "invalidation.validate", "invalidation event exceeds 4096 bytes")
	}
	return event, nil
}

func validateInvalidationValue(value any, depth int, nodes *int) error {
	*nodes++
	if *nodes > 32 {
		return forgeError(CodeLimit, "invalidation.validate", "query-key fragment has too many nodes")
	}
	switch value := value.(type) {
	case nil, bool, json.Number:
		return nil
	case string:
		if len([]byte(value)) > 128 {
			return forgeError(CodeLimit, "invalidation.validate", "query-key string exceeds 128 bytes")
		}
		return nil
	case []any:
		if depth >= 3 {
			return forgeError(CodeLimit, "invalidation.validate", "query-key nesting exceeds 3 levels")
		}
		if len(value) > 16 {
			return forgeError(CodeLimit, "invalidation.validate", "query-key array has too many items")
		}
		for _, item := range value {
			if err := validateInvalidationValue(item, depth+1, nodes); err != nil {
				return err
			}
		}
		return nil
	case map[string]any:
		if depth >= 3 {
			return forgeError(CodeLimit, "invalidation.validate", "query-key nesting exceeds 3 levels")
		}
		if len(value) > 16 {
			return forgeError(CodeLimit, "invalidation.validate", "query-key object has too many items")
		}
		for key, item := range value {
			if len([]byte(key)) > 64 {
				return forgeError(CodeLimit, "invalidation.validate", "query-key object key exceeds 64 bytes")
			}
			if err := validateInvalidationValue(item, depth+1, nodes); err != nil {
				return err
			}
		}
		return nil
	default:
		return forgeError(CodeInvalid, "invalidation.validate", "query-key parts must be JSON values")
	}
}

File: src/invalidation.rs
//! Transport-neutral, lossy hints that tell a client to re-read authoritative state.
//!
//! This is a data contract, not a ninth Forge primitive or a durable change feed. Encode an
//! event and deliver it through application-owned SSE, WebSocket, push, or Forge pub/sub.

use std::collections::HashSet;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{ForgeError, Result};

pub const INVALIDATION_SCHEMA_VERSION: u32 = 1;
pub const MAX_INVALIDATION_BYTES: usize = 4096;
pub const MAX_INVALIDATION_TAGS: usize = 32;
pub const MAX_INVALIDATION_QUERY_KEYS: usize = 32;
pub const MAX_INVALIDATION_TARGETS: usize = 64;
pub const MAX_INVALIDATION_TAG_BYTES: usize = 128;
pub const MAX_INVALIDATION_REVISION_BYTES: usize = 256;
pub const MAX_QUERY_KEY_PARTS: usize = 8;
pub const MAX_QUERY_KEY_DEPTH: usize = 3;
pub const MAX_QUERY_KEY_NODES: usize = 32;
pub const MAX_QUERY_KEY_CONTAINER_ITEMS: usize = 16;
pub const MAX_QUERY_KEY_STRING_BYTES: usize = 128;
pub const MAX_QUERY_KEY_OBJECT_KEY_BYTES: usize = 64;

/// The canonical JSON Schema for invalidation event version 1.
pub const INVALIDATION_SCHEMA_JSON: &str = include_str!("../contract/invalidation-v1.schema.json");

/// A bounded hint that names coarse application resources or query-key fragments.
///
/// Unknown JSON fields are ignored while decoding version 1. The event contains no state
/// payload: a receiver must refetch or advance its durable application cursor.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InvalidationEvent {
    pub schema_version: u32,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub query_keys: Vec<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
}

impl InvalidationEvent {
    pub fn new(
        tags: Vec<String>,
        query_keys: Vec<Vec<Value>>,
        revision: Option<String>,
    ) -> Result<Self> {
        let event = Self {
            schema_version: INVALIDATION_SCHEMA_VERSION,
            tags,
            query_keys,
            revision,
        };
        event.validate()?;
        Ok(event)
    }

    /// Decode a bounded version-1 event. Unknown additive fields are discarded.
    pub fn decode(encoded: &[u8]) -> Result<Self> {
        if encoded.len() > MAX_INVALIDATION_BYTES {
            return Err(ForgeError::limit("invalidation event exceeds 4096 bytes"));
        }
        let event: Self = serde_json::from_slice(encoded)
            .map_err(|_| ForgeError::invalid("invalidation event must be valid JSON"))?;
        event.validate()?;
        Ok(event)
    }

    /// Validate and encode the normalized version-1 event.
    pub fn encode(&self) -> Result<Vec<u8>> {
        self.validate()?;
        let encoded = serde_json::to_vec(self)
            .map_err(|_| ForgeError::invalid("invalidation event cannot be encoded"))?;
        if encoded.len() > MAX_INVALIDATION_BYTES {
            return Err(ForgeError::limit("invalidation event exceeds 4096 bytes"));
        }
        Ok(encoded)
    }

    pub fn validate(&self) -> Result<()> {
        if self.schema_version != INVALIDATION_SCHEMA_VERSION {
            return Err(ForgeError::invalid(
                "unsupported invalidation schema version",
            ));
        }
        if self.tags.is_empty() && self.query_keys.is_empty() {
            return Err(ForgeError::invalid(
                "invalidation event requires a tag or query-key fragment",
            ));
        }
        if self.tags.len() > MAX_INVALIDATION_TAGS {
            return Err(ForgeError::limit("invalidation event has too many tags"));
        }
        if self.query_keys.len() > MAX_INVALIDATION_QUERY_KEYS
            || self.tags.len() + self.query_keys.len() > MAX_INVALIDATION_TARGETS
        {
            return Err(ForgeError::limit("invalidation event has too many targets"));
        }

        let mut unique_tags = HashSet::with_capacity(self.tags.len());
        for tag in &self.tags {
            if tag.is_empty() || tag.len() > MAX_INVALIDATION_TAG_BYTES {
                return Err(ForgeError::invalid(
                    "invalidation tags must be 1..=128 UTF-8 bytes",
                ));
            }
            if !unique_tags.insert(tag) {
                return Err(ForgeError::invalid("invalidation tags must be unique"));
            }
        }

        for query_key in &self.query_keys {
            if query_key.is_empty() || query_key.len() > MAX_QUERY_KEY_PARTS {
                return Err(ForgeError::invalid(
                    "query-key fragments must contain 1..=8 parts",
                ));
            }
            let mut nodes = 0;
            for part in query_key {
                validate_query_key_value(part, 1, &mut nodes)?;
            }
        }

        if self
            .revision
            .as_ref()
            .is_some_and(|value| value.is_empty() || value.len() > MAX_INVALIDATION_REVISION_BYTES)
        {
            return Err(ForgeError::invalid(
                "invalidation revision must be 1..=256 UTF-8 bytes",
            ));
        }

        let encoded = serde_json::to_vec(self)
            .map_err(|_| ForgeError::invalid("invalidation event cannot be encoded"))?;
        if encoded.len() > MAX_INVALIDATION_BYTES {
            return Err(ForgeError::limit("invalidation event exceeds 4096 bytes"));
        }
        Ok(())
    }
}

fn validate_query_key_value(value: &Value, depth: usize, nodes: &mut usize) -> Result<()> {
    *nodes += 1;
    if *nodes > MAX_QUERY_KEY_NODES {
        return Err(ForgeError::limit("query-key fragment has too many nodes"));
    }
    match value {
        Value::Null | Value::Bool(_) | Value::Number(_) => Ok(()),
        Value::String(value) if value.len() <= MAX_QUERY_KEY_STRING_BYTES => Ok(()),
        Value::String(_) => Err(ForgeError::limit("query-key string exceeds 128 bytes")),
        Value::Array(values) => {
            validate_container(depth, values.len())?;
            for value in values {
                validate_query_key_value(value, depth + 1, nodes)?;
            }
            Ok(())
        }
        Value::Object(values) => {
            validate_container(depth, values.len())?;
            for (key, value) in values {
                if key.len() > MAX_QUERY_KEY_OBJECT_KEY_BYTES {
                    return Err(ForgeError::limit("query-key object key exceeds 64 bytes"));
                }
                validate_query_key_value(value, depth + 1, nodes)?;
            }
            Ok(())
        }
    }
}

fn validate_container(depth: usize, items: usize) -> Result<()> {
    if depth >= MAX_QUERY_KEY_DEPTH {
        return Err(ForgeError::limit("query-key nesting exceeds 3 levels"));
    }
    if items > MAX_QUERY_KEY_CONTAINER_ITEMS {
        return Err(ForgeError::limit("query-key container has too many items"));
    }
    Ok(())
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::indexing_slicing)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn round_trip_discards_unknown_additive_fields() {
        let event = InvalidationEvent::decode(
            br#"{"schema_version":1,"tags":["links"],"query_keys":[["links",{"owner":"u1"}]],"revision":"42","future":{"ignored":true}}"#,
        )
        .unwrap();
        assert_eq!(event.tags, ["links"]);
        assert!(
            !String::from_utf8(event.encode().unwrap())
                .unwrap()
                .contains("future")
        );
    }

    #[test]
    fn rejects_state_payloads_by_size_and_unbounded_shapes() {
        let oversized = vec![b'x'; MAX_INVALIDATION_BYTES + 1];
        assert_eq!(
            InvalidationEvent::decode(&oversized).unwrap_err().code(),
            "LIMIT"
        );
        let nested = InvalidationEvent::new(vec![], vec![vec![json!([[[["too-deep"]]]])]], None)
            .unwrap_err();
        assert_eq!(nested.code(), "LIMIT");
    }

    #[test]
    fn rejects_empty_or_duplicate_targets() {
        assert_eq!(
            InvalidationEvent::new(vec![], vec![], None)
                .unwrap_err()
                .code(),
            "INVALID"
        );
        assert_eq!(
            InvalidationEvent::new(vec!["links".into(), "links".into()], vec![], None)
                .unwrap_err()
                .code(),
            "INVALID"
        );
    }
}
